examples/cis2-multi/src/lib.rs

```diff
@@ -440,45 +686,215 @@ fn contract_transfer(
+    // Calculate the message hash.
+    let message_hash =
+        crypto_primitives.hash_sha2_256(&[&msg_prepend[0..40], &message_bytes].concat()).0;
+
+    Ok(message_hash)
+}
+
+/// Verify an ed25519 signature and allow the transfer of tokens or update of an
+/// operator.
+///
+/// In case of a `transfer` action:
+/// Logs a `Transfer` event and invokes a receive hook function for the
+/// transfer.
+///
+/// In case of a `updateOperator` action:
+/// Logs an `UpdateOperator` event.
+///
+/// It rejects if:
+/// - It fails to parse the parameter.
+/// - A different nonce is expected.
+/// - The signature was intended for a different contract.
+/// - The signature was intended for a different `entry_point`.
+/// - The signature is expired.
+/// - The signature can not be validated.
+/// - Fails to log event.
+/// - In case of a `transfer` action: it fails to be executed, which could be
+///   if:
+///     - The `token_id` does not exist.
+///     - The token is not owned by the `from` address.
+///     - The receive hook function call rejects.
+#[receive(
+    contract = "cis2_multi",
+    name = "permit",
+    parameter = "PermitParam",
+    crypto_primitives,
+    mutable,
+    enable_logger
+)]
+fn contract_permit(
+    ctx: &ReceiveContext,
+    host: &mut Host<State>,
+    logger: &mut impl HasLogger,
+    crypto_primitives: &impl HasCryptoPrimitives,
+) -> ContractResult<()> {
+    // Parse the parameter.
+    let param: PermitParam = ctx.parameter_cursor().get()?;
+
+    // Update the nonce.
+    let mut entry = host.state_mut().nonces_registry.entry(param.signer).or_insert_with(|| 0);
+
+    // Get the current nonce.
+    let nonce = *entry;
+    // Bump nonce.
+    *entry += 1;
+    drop(entry);
+
+    let message = param.message;
+
+    // Check the nonce to prevent replay attacks.
+    ensure_eq!(message.nonce, nonce, CustomContractError::NonceMismatch.into());
+
+    // Check that the signature was intended for this contract.
+    ensure_eq!(
+        message.contract_address,
+        ctx.self_address(),
+        CustomContractError::WrongContract.into()
+    );
+
+    // Check signature is not expired.
+    ensure!(message.timestamp > ctx.metadata().slot_time(), CustomContractError::Expired.into());
+
+    let message_hash = contract_view_message_hash(ctx, host, crypto_primitives)?;
+
+    // Check signature.
+    let valid_signature =
+        host.check_account_signature(param.signer, &param.signature, &message_hash)?;
+    ensure!(valid_signature, CustomContractError::WrongSignature.into());
+
+    if message.entry_point.as_entrypoint_name() == EntrypointName::new_unchecked("transfer") {
+        // Transfer the tokens.
+
+        let TransferParams(transfers): TransferParameter = from_bytes(&message.payload)?;
+
+        for transfer_struct in transfers {
+            ensure!(
+                transfer_struct.from.matches_account(&param.signer),
+                ContractError::Unauthorized
+            );
```

Comment:
I don't have a very strong explanation for why we should or should not allow operators to sign transactions as well. It should be consistent in the CIS3 standard examples to avoid confusion.

In general, an operator of your account is usually a smart contract (I don't see any good reason why someone should set an account as their operator since this gives unrestricted control over all your assets to a third party). Since a smart contract cannot generate a signature on Concordium, it seems to not really make a big difference if we decide to allow or not allow operators in the logic of the CIS3 standard to generate such signatures.

Reply:
> Since a smart contract cannot generate a signature on Concordium, it seems to not really make a big difference

Yes, good point.

> In general, an operator of your account is usually a smart contract (I don't see any good reason why someone should set an account as their operator since this gives unrestricted control over all your token assets to a third party). 

Yes, I argee. One use case could be that the user makes other accounts owned by this (that same) user into operators. Not clear why it is useful, so indeed, not using operators does not restrict the functionality too much.

Reply:
Ok, let's discuss it separately and if we decide to use operators in `permit` for transfer, it will be a separate PR changing this behaviour in all sponsored transaction examples.